components/background_hang_monitor/sampler.rs

```diff
@@ -0,0 +1,337 @@
+            thread_id
+        );
+    }
+}
+
+#[cfg(target_os = "windows")]
+#[allow(unsafe_code)]
+unsafe fn resume_thread(thread_id: MonitoredThreadId) {
+    // TODO: use winapi::um::processthreadsapi::ResumeThread
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+fn resume_thread(thread_id: MonitoredThreadId) {
+    SAMPLEE_MAY_RESUME.store(true, Ordering::SeqCst);
+    let now = Instant::now();
+    // Wait until the samplee has indeed resumed,
+    // signaled by setting the flag back to false.
+    loop {
+        if !SAMPLEE_MAY_RESUME.load(Ordering::SeqCst) {
+            break;
+        }
+        if now.elapsed().as_secs() > 5 {
+            // The samplee still hasn't resumed, assume failure to resume.
+            warn!(
+                "Background-hang-monitor failed to resume thread: {:?}",
+                thread_id
+            );
+            break;
+        }
+        thread::yield_now();
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe fn frame_pointer_stack_walk(regs: Registers) -> NativeStack {
+    // Note: this will only work with code build with:
+    // --dev,
+    // or --with-frame-pointer.
+    let mut native_stack = NativeStack::new();
+    let pc = regs.instruction_ptr as *mut std::ffi::c_void;
+    let stack = regs.stack_ptr as *mut std::ffi::c_void;
+    native_stack.process_register(pc, stack);
+    let mut current = regs.frame_ptr as *mut *mut std::ffi::c_void;
+    let mut frame_count = 1;
+    while !current.is_null() && frame_count < MAX_NATIVE_FRAMES {
+        let next = *current as *mut *mut std::ffi::c_void;
+        let pc = current.add(1);
+        let stack = current.add(2);
+        native_stack.process_register(*pc, *stack);
+        frame_count = frame_count + 1;
+        current = next;
+    }
+    native_stack
+}
+
+fn symbolize_backtrace(native_stack: NativeStack) -> HangProfile {
+    let mut profile = HangProfile {
+        backtrace: Vec::new(),
+    };
+    for ip in native_stack.instruction_ptrs.iter().rev() {
+        if ip.is_null() {
+            continue;
+        }
+        backtrace::resolve(*ip, |symbol| {
+            let mut hang_symbol = HangProfileSymbol {
+                name: None,
+                filename: None,
+                lineno: None,
+            };
+            if let Some(symbol_name) = symbol.name() {
+                let bytes = symbol_name.as_bytes();
+                hang_symbol.name = Some(String::from_utf8_lossy(&bytes).to_string());
+            }
+            if let Some(filename) = symbol.filename() {
+                let name = filename.to_string_lossy();
+                hang_symbol.filename = Some(name.to_string());
+            }
+            hang_symbol.lineno = symbol.lineno();
+            profile.backtrace.push(hang_symbol);
+        });
+    }
+    profile
+}
+
+#[allow(unsafe_code)]
+pub unsafe fn suspend_and_sample_thread(thread_id: MonitoredThreadId) -> Option<HangProfile> {
+    // Warning: The "critical section" begins here.
```

Comment:
I think panicking is going to need special handling inside the critical section.
The default panic handler will allocate memory and try to print a backtrace, which we definitely don't want to do in the critical section.

I think the best solution would be to use `std::panic::set_hook` to register a custom panic handler which unconditionally aborts. The old panic handler can be retrieved with `std::panic::take_hook`, and restored once the critical section has ended.

I don't think that using `std::panic::catch_unwind` will work, since it allocates a `Box` to hold the error.